### PR TITLE
crimson/os/seastore/cbj: fix a potential overflow bug on segment_seq

### PR DESCRIPTION
**Pull request by myoungwon**
CircularBoundedJournal is similar to circular queue, so segment_seq increases after rolling. However, current implementation always increases segment_seq_t when rolling occurs, resulting in the overflow if segment_seq_t hits MAX_SEG_SEQ.

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
Is 32-bit `segment_seq_t` not enough ?

Assuming the segment/cbj size is at the level of 100 MB, 32-bit can support lifetime writes to an SSD at the level of EB.

Seems to me tolerating segment_seq_t overflow by resetting it to 0 is not enough, any seastore logic depends on comparing its absolute value will break (e.g. comparing `journal_seq_t`, etc).

**Comment by myoungwon** [comment]
> Is 32-bit segment_seq_t not enough ?
> Assuming the segment/cbj size is at the level of 100 MB, 32-bit can support lifetime writes to an SSD at the level of EB.

I think, for now, the 32-bit segment_seq_t is probably enough in normal situation. But, if we consider a case where the same operations are repetitive like write(0, 4096) - remove(0, 4096) - write(0, 4096) - remove(0, 4096)...,  the journal can run into the maximum size even though the RBM is not full. This is kind of the extreme case, but it can happen. 

> Seems to me tolerating segment_seq_t overflow by resetting it to 0 is not enough, any seastore logic depends on comparing its absolute value will break (e.g. comparing journal_seq_t, etc).

Right. Hm.. How can we overcome this? For now, at least, I think we need to add an assertion if the segment_seq_t reaches the maximum value before a solution is introduced.

**Comment by cyx1231st** [comment]
> But, if we consider a case where the same operations are repetitive like write(0, 4096) - remove(0, 4096) - write(0, 4096) - remove(0, 4096)..., the journal can run into the maximum size even though the RBM is not full.

By EBs I think it is approximately the seastore lifetime writes to the journaling area of a device, not related to whether it is full.

> Right. Hm.. How can we overcome this? For now, at least, I think we need to add an assertion if the segment_seq_t reaches the maximum value before a solution is introduced.

Yeah, I think currently adding `ceph_assert` when increment the seg-seq should work, this issue also applies to the Segment backend. If the 32-bit sequence proves to be not enough, 64-bit should suffice.

**Comment by myoungwon** [comment]
Regarding ssd’s lifetime,
I asked samsung firmware engineer about ssd’s lifetime. Consequently, it is difficult to predict ssd’s lifetime using the write count in int32. As you may know, although the ssd worn out if it hits a limited block erase or write count, it does not mean that ssd must be unavailable if the count exceeds over the threshold because it depends on ssd’s characteristic and lifespan technique such as wear-leveling—spreading write to multiple blocks.

Therefore, at least, it seems that we need to change the segment_seq_t to 64 bit value and adding assert for now even thought potential problem remains to solve in the future. Am i missing something?

**Comment by cyx1231st** [comment]
> Therefore, at least, it seems that we need to change the segment_seq_t to 64 bit value and adding assert for now even thought potential problem remains to solve in the future. Am i missing something?

Looks reasonable to me. 64-bit should be good enough considering the journaling area won't be too small.

**Comment by myoungwon** [comment]
@cyx1231st Updated.

File: src/crimson/os/seastore/journal/circular_journal_space.cc [hunk 1]
```diff
@@ -42,8 +42,9 @@ CircularJournalSpace::roll_ertr::future<> CircularJournalSpace::roll() {
     get_records_start(),
     get_device_id());
   auto seq = get_written_to();
+  assert(++seq.segment_seq < MAX_SEG_SEQ);
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/journal/circular_journal_space.cc]: I'm sorry. Fixed.

**Review by liu-chunmei — APPROVED** [review]

**Review by cyx1231st — APPROVED** [review]

**Pull request merged**